src/idom/utils.py

```diff
@@ -63,80 +63,130 @@ def html_to_vdom(source: str, *transforms: _ModelTransform) -> Dict[str, Any]:
+    if has_root_node:
+        root_node = nodes[0]
+    else:
+        root_node = etree.Element("div", None, None)
+        for child in nodes:
+            root_node.append(child)
+
+    # Convert the lxml node to a VDOM dict
+    vdom = _etree_to_vdom(root_node, *transforms)
+
+    # Change the artificially created root node to a React Fragment, instead of a div
+    if not has_root_node:
+        vdom["tagName"] = ""
+
+    return vdom
+
+
+def _etree_to_vdom(node: etree._Element, *transforms: _ModelTransform) -> Dict:
+    """Recusively transform an lxml etree node into a DOM model
+    Parameters:
+        source:
+            The ``lxml.etree._Element`` node
+        transforms:
+            Functions of the form ``transform(old) -> new`` where ``old`` is a VDOM
+            dictionary which will be replaced by ``new``. For example, you could use a
+            transform function to add highlighting to a ``<code/>`` block.
     """
-
-    def model(self) -> Dict[str, Any]:
-        """Get the current state of parsed VDOM model"""
-        return self._node_stack[0]
-
-    def feed(self, data: str) -> None:
-        """Feed in HTML that will update the :meth:`HtmlParser.model`"""
-        self._node_stack.append(self._make_vdom("div", {}))
-        super().feed(data)
-
-    def reset(self) -> None:
-        """Reset the state of the parser"""
-        self._node_stack: List[Dict[str, Any]] = []
-        super().reset()
-
-    def handle_starttag(self, tag: str, attrs: List[Tuple[str, Optional[str]]]) -> None:
-        new = self._make_vdom(tag, dict(attrs))
-        current = self._node_stack[-1]
-        current["children"].append(new)
-        self._node_stack.append(new)
-
-    def handle_endtag(self, tag: str) -> None:
-        del self._node_stack[-1]
-
-    def handle_data(self, data: str) -> None:
-        self._node_stack[-1]["children"].append(data)
-
-    @staticmethod
-    def _make_vdom(tag: str, attrs: Dict[str, Any]) -> Dict[str, Any]:
-        if "style" in attrs:
-            style = attrs["style"]
-            if isinstance(style, str):
-                style_dict = {}
-                for k, v in (part.split(":", 1) for part in style.split(";") if part):
-                    title_case_key = k.title().replace("-", "")
-                    camel_case_key = title_case_key[:1].lower() + title_case_key[1:]
-                    style_dict[camel_case_key] = v
-                attrs["style"] = style_dict
-        return {"tagName": tag, "attributes": attrs, "children": []}
+    if not isinstance(node, etree._Element):
+        raise TypeError(f"Encountered unsupported type {type(node)} from {node}")
+
+    # Convert the lxml node to a VDOM dict.
+    vdom = {
+        "tagName": node.tag,
+        "children": _generate_vdom_children(node, transforms),
+        "attributes": dict(node.items()),
+        "eventHandlers": {},
+        "importSource": {},
+        "key": "",
+        "error": "",
+    }
```

Comment:
If the `hasattr(idom.html, tagName)`, let's use the constructors from that module. This will give useful errors for elements which are not allowed to have children as well has handle any extra logic (e.g. the `script` element). For elements not in the `idom.html` module, you'll only need to declare the `tagName`,  `children`, `attributes`, and `key` fields - the others cannot be defined via static HTML at the moment.

For the `key` of an element, you'll want to pop that off the attributes dict and assign that to the VDOM dict, or hand it to an `html` constructor, separately - this was not properly handled in the old version either.